launchlib/processors.go

```diff
@@ -0,0 +1,148 @@
+// Copyright 2016 Palantir Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package launchlib
+
+import (
+	"bufio"
+	"bytes"
+	"io"
+	"io/fs"
+	"math"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	selfCGroup    = "/proc/self/cgroup"
+	selfMountinfo = "/proc/self/mountinfo"
+	cpuGroupName  = "cpu"
+	cpuSharesName = "cpu.shares"
+)
+
+type ProcessorCounter interface {
+	ProcessorCount() (uint, error)
+}
+
+var DefaultCGroupV1ProcessorCounter = CGroupV1ProcessorCounter{
+	fs: os.DirFS("/"),
+}
+
+type CGroupV1ProcessorCounter struct {
+	fs fs.FS
+}
+
+func NewCGroupV1ProcessorCounter(filesystem fs.FS) ProcessorCounter {
+	return CGroupV1ProcessorCounter{fs: filesystem}
+}
+
+func (c CGroupV1ProcessorCounter) ProcessorCount() (uint, error) {
```

Comment:
When I was doing this for memory in gradle, there's a case where cgroups is installed but there is no limit for the current container/environment. An example of this is Circle machine executors (I wonder if PCloud is like this too?). Under cgroupsv1 you get [a really big value](https://github.com/CRogers/gradle/commit/3fb0f7d6548aacf662d863b34a09b88c06a11e18#diff-53102e5845be037fa8056d1330c50c4c74d69a5c9b9414f31c8c40fb2a75bb93R30-R32) (2^60 or something) which means "no limit" but you can end up thinking there's more memory than there is. Is this a problem as well for cpu shares implementation?

Reply:
The issue I want to avoid here is thinking we have 2^63 cpu shares, so tell java we have 2^63/1024 = 2^54 processors then witchcraft tries to make a thread pool with 8 * 2^54 = 2^57 threads.

Reply:
As best as I can tell, groups get `1024` as their [cpu shares by default](https://kernel.googlesource.com/pub/scm/linux/kernel/git/glommer/memcg/+/cpu_stat/Documentation/cgroups/cpu.txt), so if someone has a very large value here, it would've had to be set manually. Maybe a safeguard here would be to detect the hosts number of processors? So something like:

```
max(2, min(runtime.NumCPUs(), floor(cpu.shares / 1024)))
```

Reply:
https://github.com/uber-go/automaxprocs provides some similar functionality on the golang side for setting `GOMAXPROCS`

Reply:
fwiw I checked on an internal machine Circle executor and the cpu shares does appear to be 1024 and not some huge number (where as the memory limit is 2^63).